unit/java_bytecode/java_bytecode_convert_method/convert_invoke_dynamic.cpp

```diff
@@ -0,0 +1,674 @@
+#include <testing-utils/require_expr.h>
+#include <testing-utils/require_type.h>
+#include <testing-utils/run_test_with_compilers.h>
+#include <testing-utils/require_symbol.h>
+#include <util/expr_iterator.h>
+#include <goto-programs/class_hierarchy.h>
+
+struct lambda_assignment_test_datat
+{
+  irep_idt lambda_interface;
+  std::string lambda_interface_method_descriptor;
+  irep_idt lambda_function_id;
+
+  std::vector<exprt> expected_params;
+  bool should_return_value = true;
+};
+
+/// Verifies for a given function that contains:
+///  {lambda_variable_id} = invokedynamic method_handle_index
+/// code looks
+/// like:
+///  tmp_object_symbol = java_new(lambda_tmp_type)
+///  {lambda_variable_id} = (cast)tmp_object_symbol
+/// and the type lambda_tmp_type:
+///  implements {lambda_interface}
+///  has method {lambda_interface_method_descriptor}
+/// and the method body looks like
+///  function_call_exprt(
+///    should_return_value ? symbol_exprt : nil_exprt,
+///    {lambda_function_id},
+///    {expected_params})
+/// \param symbol_table: The loaded symbol table
+/// \param instructions: The instructions of the method that calls invokedynamic
+/// \param test_data: The parameters for the test
+void validate_lambda_assignment(
+  const symbol_tablet &symbol_table,
+  const std::vector<codet> &instructions,
+  const lambda_assignment_test_datat &test_data,
+  const require_goto_statements::pointer_assignment_locationt
+  &lambda_assignment)
+{
+  const typecast_exprt &rhs_value = require_expr::require_typecast(
+    lambda_assignment.non_null_assignments[0].rhs());
+
+  const symbol_exprt &rhs_symbol =
+    require_expr::require_symbol(rhs_value.op0());
+
+  const irep_idt &tmp_object_symbol = rhs_symbol.get_identifier();
+
+  const auto tmp_object_assignments =
+    require_goto_statements::find_pointer_assignments(
+      tmp_object_symbol, instructions);
+
+  REQUIRE(tmp_object_assignments.non_null_assignments.size() == 1);
+  REQUIRE_FALSE(tmp_object_assignments.null_assignment.has_value());
+
+  const side_effect_exprt &side_effect_expr =
+    require_expr::require_side_effect_expr(
+      tmp_object_assignments.non_null_assignments[0].rhs(), ID_java_new);
+
+  const pointer_typet &lambda_temp_type =
+    require_type::require_pointer(side_effect_expr.type(), {});
+
+  const symbol_typet &lambda_implementor_type =
+    require_type::require_symbol(lambda_temp_type.subtype());
+
+  const irep_idt &tmp_class_identifier =
+    lambda_implementor_type.get_identifier();
+
+  const symbolt &lambda_implementor_type_symbol =
+    require_symbol::require_symbol_exists(symbol_table, tmp_class_identifier);
+
+  REQUIRE(lambda_implementor_type_symbol.is_type);
+  const class_typet &tmp_lambda_class_type =
+    require_type::require_complete_class(lambda_implementor_type_symbol.type);
+
+  REQUIRE(tmp_lambda_class_type.has_base(test_data.lambda_interface));
+  REQUIRE(tmp_lambda_class_type.has_base("java::java.lang.Object"));
```

Comment:
Nit: The `Object` is here because it is a superclass correct? It should appear first in the list of bases, so consider swapping the require statements for readability.